File: packages/sdkwork-utils-java/src/main/java/com/sdkwork/utils/CompareUtils.java
package com.sdkwork.utils;

import java.util.List;
import java.util.Map;
import java.util.Objects;

public final class CompareUtils {
    private CompareUtils() {
    }

    public static boolean deepEqual(Object left, Object right) {
        if (left == right) {
            return true;
        }
        if (left == null || right == null) {
            return false;
        }
        if (left instanceof Map<?, ?> leftMap && right instanceof Map<?, ?> rightMap) {
            if (leftMap.size() != rightMap.size()) {
                return false;
            }
            for (Map.Entry<?, ?> entry : leftMap.entrySet()) {
                Object key = entry.getKey();
                if (!rightMap.containsKey(key)) {
                    return false;
                }
                if (!deepEqual(entry.getValue(), rightMap.get(key))) {
                    return false;
                }
            }
            return true;
        }
        if (left instanceof List<?> leftList && right instanceof List<?> rightList) {
            if (leftList.size() != rightList.size()) {
                return false;
            }
            for (int index = 0; index < leftList.size(); index++) {
                if (!deepEqual(leftList.get(index), rightList.get(index))) {
                    return false;
                }
            }
            return true;
        }
        return Objects.equals(left, right);
    }

    @SuppressWarnings("unchecked")
    public static Object deepClone(Object value) {
        if (value instanceof Map<?, ?> map) {
            Map<Object, Object> cloned = new java.util.LinkedHashMap<>();
            for (Map.Entry<?, ?> entry : map.entrySet()) {
                cloned.put(entry.getKey(), deepClone(entry.getValue()));
            }
            return cloned;
        }
        if (value instanceof List<?> list) {
            List<Object> cloned = new java.util.ArrayList<>(list.size());
            for (Object item : list) {
                cloned.add(deepClone(item));
            }
            return cloned;
        }
        return value;
    }
}

File: packages/sdkwork-utils-csharp/CompareUtils.cs
namespace Sdkwork.Utils;

public static class CompareUtils
{
    public static bool DeepEqual(object? left, object? right)
    {
        if (ReferenceEquals(left, right))
        {
            return true;
        }

        if (left is null || right is null)
        {
            return false;
        }

        if (left is IDictionary<string, object?> leftMap && right is IDictionary<string, object?> rightMap)
        {
            if (leftMap.Count != rightMap.Count)
            {
                return false;
            }

            foreach (var entry in leftMap)
            {
                if (!rightMap.TryGetValue(entry.Key, out var other) || !DeepEqual(entry.Value, other))
                {
                    return false;
                }
            }

            return true;
        }

        if (left is IList<object?> leftList && right is IList<object?> rightList)
        {
            if (leftList.Count != rightList.Count)
            {
                return false;
            }

            for (var index = 0; index < leftList.Count; index++)
            {
                if (!DeepEqual(leftList[index], rightList[index]))
                {
                    return false;
                }
            }

            return true;
        }

        return Equals(left, right);
    }

    public static object? DeepClone(object? value)
    {
        return value switch
        {
            IDictionary<string, object?> map => map.ToDictionary(entry => entry.Key, entry => DeepClone(entry.Value)),
            IList<object?> list => list.Select(DeepClone).ToList(),
            _ => value
        };
    }
}
